### ISSUE TITLE
addMethod(Schema) not working in TypeScript files

### ISSUE THREAD

**Issue by csantos1113** [jquense/yup#2068]
FROM the docs https://github.com/jquense/yup#addmethodschematype-schema-name-string-method--schema-void

> If you want to add a method to ALL schema types, extend the abstract base class: Schema
>
> ```js
> import { addMethod, Schema } from 'yup';
> 
> addMethod(Schema, 'myMethod', ...)
> ```

It doesn't work in `.ts` files

---

_Originally posted by @csantos-nydig in https://github.com/jquense/yup/issues/2058#issuecomment-1638215239_

I'm not sure you looked at the [codesandbox.io/s/mixed-yup1-0-2-forked-l56m8r](https://codesandbox.io/s/mixed-yup1-0-2-forked-l56m8r), but let me add an image here to show you where the TypeScript errors shows up:

<img width="777" alt="Screenshot 2023-07-17 at 8 35 04 AM" src="https://github.com/jquense/yup/assets/87080313/17ccdb5f-eaad-42d1-9b45-a837f815a35b">

---

I also cloned `yup` repository locally, and pasted in my code in one of the ts files: the typescript error is the same

<img width="536" alt="image" src="https://github.com/jquense/yup/assets/87080313/ff6606ab-617e-41ca-b8b7-84ace6ff413c">

---

I see you have unit tests for `addMethod` https://github.com/jquense/yup/blob/master/test/yup.js#L206-L211, but that file is written in JavaScript, so the TypeScript error doesn't show up

but the TS error appears if you move this simple line from `.js` to `.ts`:

```
addMethod(Schema, 'foo', () => 'here');
```

<img width="1499" alt="image" src="https://github.com/jquense/yup/assets/87080313/9b5adf6b-197f-47fa-807d-0d833554c100">

---

I don't think it has anything to do with the [`declare` statement](https://github.com/jquense/yup#extending-built-in-schema-with-new-methods), which is actually working: see the green box in the screenshot above, and it also works fine when executed.

<img width="729" alt="image" src="https://github.com/jquense/yup/assets/87080313/44f8976f-9413-40b5-b33a-b8c19e93b6b4">

So **the TS error is when calling `addMethod` on `Schema` not when using the added method.** 👇

<img width="922" alt="image" src="https://github.com/jquense/yup/assets/87080313/91be7e55-6eae-4a53-ae98-3b87828c8b61">

**Comment by StevenVerbiest** [comment]
You are using the `Schema` type as a value? 

If you want to add your method to the string schema, you can do this:

```
declare module "yup" {
  interface StringSchema {
    forAll(): StringSchema;
  }
}

addMethod(string, "forAll", function forAll() {
  return this.transform(() => {
    return "just for testing";
  });
});

const stringSchema = string().forAll();
```

And if you want to add it to the mixed schema:

```
declare module "yup" {
  interface MixedSchema {
    forAll(): MixedSchema;
  }
}

addMethod(mixed, "forAll", function forAll() {
  return this.transform(() => {
    return "just for testing";
  });
});

const mixedSchema = mixed().forAll();
```

**Comment by csantos1113** [comment]
@StevenVerbiest thanks for your reply. No, from the docs I want to add a method to ALL types:

https://github.com/jquense/yup#addmethodschematype-schema-name-string-method--schema-void

> If you want to add a method to ALL schema types, extend the abstract base class: `Schema`

**Comment by csantos1113** [comment]
This is fixed in [v1.3.3](https://github.com/jquense/yup/blob/master/CHANGELOG.md#133-2023-12-14) via f921fe69a2d6ecc6e7d0101d2bd81148dfe83e64
